### DISCUSSION TITLE
How do I make sure my calculations are correct when using type double?

### DISCUSSION THREAD

**Discussion by ChenKe404** [CGAL/cgal#8126] [Q&A]
For example, If I want to know if a point is on a segment(like use "bounded_side"), when the number type is double, the result maybe is wrong.
If I specialize the template function "CGAL::compare" like this(from network):

```c++
double tolerance = 1e-4;   // two numbers are equal if their difference is within this range

template <>
inline
    typename CGAL::Real_embeddable_traits<typename CGAL::Coercion_traits<double,double>::Type>
    ::Compare::result_type
    CGAL::compare(const double& a,const double& b)
{
    if((a - b) < -tolerance){
        return SMALLER;
    }
    else if((a - b) > tolerance){
        return LARGER;
    }
    else{
        return EQUAL;
    }
}
```

OK, the "bounded_side" has a correct result now.
But I dont know that `CGAL::intersection`, `CGAL::join` ... , these functions can also get the correct results I want.
And I can't find anything in the documentation about: "How to specialize those templates" and "which templates should be specialized".
Someone help?
Sorry about my english.

**Comment by MaelRL** [comment]
You likely shouldn't be at this level, but should rather choose and use a kernel that will provide the exact predicates that  you need.

You can have a look at:
https://www.cgal.org/exact.html
https://doc.cgal.org/latest/Kernel_23/index.html
